### ISSUE TITLE
EC2 security group definitions don't really work

### ISSUE THREAD

**Issue by nh2** [NixOS/nixops#684]
I define a setup like

```nix
      resources.ec2SecurityGroups.mygroup = {
        accessKeyId = "myaccesskeyid";
        region = "eu-central-1";
        name = "mygroup";
        description = "mygroup";
        rules = [
          { fromPort = 22; toPort = 22; sourceIp = "0.0.0.0/0"; }
          { fromPort = 80; toPort = 80; sourceIp = "0.0.0.0/0"; }
        ];
      };
```

It creates the entries fine, but if I delete one (e.g. the one for port 80) from the AWS console and then run this again, it doesn't recreate the entry I deleted.

I believe this is because the sqlite database simply stores what it believes the remote state is, and doesn't check it.

Similarly, when I create an entry manually (e.g. one for port 1234), nixops doesn't make sure that it's gone when I run it.

Not a very declarative deployment behaviour in this case.

**Comment by AmineChikhaoui** [comment]
@nh2 Could you try with --check and add --debug if there are issues. There are some issues that we discovered as part of https://github.com/NixOS/nixops/pull/670 so might be related.

**Comment by nh2** [comment]
With `--check` I get an immediate error:

`error: 'EC2SecurityGroupState' object has no attribute 'Missing'`

**Comment by nh2** [comment]
Which looks like a thing in that linked PR: https://github.com/NixOS/nixops/pull/670/files#diff-8ae8bc557ff80acd1b94e9092a487726R136

**Comment by AmineChikhaoui** [comment]
ok thanks @nh2 . I'll have a look at this case as part of the fixes in #670 .

**Comment by AmineChikhaoui** [comment]
@nh2 Can you double check if #670 fixed the issue ? I have tested removing a rule from the cli and then redeploying with `--check` and it should work now.

**Comment by domenkozar** [comment]
@nh2 ping

**Comment by nh2** [comment]
I'll try it out when I have time -- you can also close and I will reopen if it doesn't work for me if you like.

**Comment by eqyiel** [comment]
I don't have time to come up with a minimal example right now, but it looks like there is still some kind of problem bootstrapping security groups.

This may be just an issue with the docs though: https://github.com/NixOS/nixops/blob/f790182a819f7fd534a88c160a11305f0b925d32/nix/ec2-security-group.nix#L37

Without explicitly setting security group IDs, I'm seeing a lot of:

```
node............................> got (possibly transient) EC2 error code 'InvalidParameterValue': Value () for parameter groupId is invalid. The value cannot be empty. retrying...
node............................> got (possibly transient) EC2 error code 'InvalidParameterValue': Value () for parameter groupId is invalid. The value cannot be empty. retrying...
node............................> got (possibly transient) EC2 error code 'InvalidParameterValue': Value () for parameter groupId is invalid. The value cannot be empty. retrying...
node............................> got (possibly transient) EC2 error code 'InvalidParameterValue': Value () for parameter groupId is invalid. The value cannot be empty. retrying...
```

**Comment by weskerfoot** [comment]
@eqyiel I'm getting the same issue. Do we have to explicitly set the groupId, or is this a case of NixOps not creating it before trying to reference the security group when deploying to a new VPC? I'm confused.

**Comment by nh2** [comment]
@AmineChikhaoui I've had time to test this now with the commits from #670 (I'm on tag `v1.6`).

With `--check`, it now recreates the security group:

```
nixops.......> creating EC2 security group ‘nixops’...
```

But the problem is that in contrast to the first nixops run, which outputs

```
nixops.......> creating EC2 security group ‘nixops’...
nixops.......> adding new rules to EC2 security group ‘nixops’...
```

the second step `adding new rules to EC2 security group` is not performed.

As a result, after `--check` the security group `nixops` exits, but it has no inbound or outbound rules (that means my declared ports `22` and `80` are not let through).

**Comment by nh2** [comment]
(Thus we can reopen please.)

**Comment by nh2** [comment]
OK more concretely, if I add a *new* rule entry like

```
{ protocol = "icmp"; sourceIp = "0.0.0.0/0"; typeNumber = -1; codeNumber = -1; }
```

then it's going to create that.

But if I then delete the entire security group, it will not recreate the rules inside.

In other words, it only re-runs adding rules for rules that _haven't been added at any time before_.

So I think this remains the problem that nixops remembers somewhere in its database that the rule has been added before, so it won't recreate it, even if it notices that the entire security group has been deleted.

**Comment by nh2** [comment]
I have a fix: PR #1171
